File: core/src/ru/myitschool/dcrawler/effects/AttackEffect.java
package ru.myitschool.dcrawler.effects;

import com.badlogic.gdx.graphics.Texture;
import ru.myitschool.dcrawler.entities.Entity;

/**
 * Created by Voyager on 04.06.2017.
 */
public class AttackEffect extends Effect {

    private String name = "Attack";
    private String description = "You may attack or make second move in this turn.";
    private Texture icon = new Texture("sword.png");
    private boolean used = false;
    private boolean positive = true;

    public AttackEffect(Entity entity) {
        super(entity);
    }

    @Override
    public Texture getIcon() {
        return icon;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public boolean isSkillUse() {
        return true;
    }

    @Override
    public boolean isPositive() {
        return true;
    }

    @Override
    public boolean isStackable() {
        return false;
    }

    @Override
    public int getStackSize() {
        return 1;
    }

    @Override
    public boolean isExpiring() {
        return false;
    }

    @Override
    public int getExpireTurns() {
        return 0;
    }

    @Override
    public String getType() {
        return "main.dcrawler.effect.attack";
    }

    @Override
    public void startMove() {
        if (getEntity().isMoved()) {
            getEntity().addMp(getEntity().getMpMax());
            used = true;
        }if (getEntity().isMoved()){
            setDelete(true);
            Entity.updateSkills();
        }
    }

    @Override
    public void endMove() {
    }

    @Override
    public int countMp(boolean withMovement) {
        if (getEntity().isMoved()&& !isDelete() && !used) { //TODO don't work if starting mp is 0
            if (!getEntity().isMovement()) {
                System.out.println("Adding MP");
                if (getEntity().getPath() != null) {
                    setHide(true);
                } else {
                    setHide(false);
                }
                return getEntity().getMpMax();
            }
        } else if (getEntity().isMoved() && !withMovement && !isDelete() && !used){
            if (!getEntity().isMovement()) {
                return getEntity().getMpMax();
            }
        } else if (isHide() && !getEntity().isMovement()) {
            setHide(false);
        }
        return 0;
    }

    @Override
    public boolean canUseSkill() {
        return !isDelete();
    }

    @Override
    public void endSkill() {
        setDelete(true);
        Entity.updateSkills();
    }

    @Override
    public void endTurn() {
        setDelete(true);
    }
}

File: core/src/com/wolg_vlad/dcrawler/effects/AttackEffect.kt
package com.wolg_vlad.dcrawler.effects

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.entities.Entity

/**
 * Created by Voyager on 04.06.2017.
 */
class AttackEffect(entity: Entity?) : Effect(entity) {
    override val name = "Attack"
    override val description = "You may attack or make second move in this turn."
    override val icon = Texture("sword.png")
    private var used = false
    override val positive: Boolean = true
    override val skillUse: Boolean = true
    override val stackable: Boolean = false
    override val stackSize: Int = 1
    override val expiring: Boolean = false
    override val expireTurns: Int = 0
    override val id: String = "main.dcrawler.effect.attack"

    override fun startMove() {
        if (entity!!.isMoved) {
            entity!!.addMp(entity!!.speed)
            used = true
        }
        if (entity!!.isMoved) {
            delete = true
            Entity.updateSkills()
        }
    }

    override fun endMove() {}
    override fun countMp(withMovement: Boolean): Int {
        if (entity != null) {
            if (entity!!.isMoved && !delete && !used) { //TODO don't work if starting mp is 0
                if (!entity!!.isMovement) {
                    println("Adding MP")
                    hide = entity?.path != null
                    return entity!!.speed
                }
            } else if (entity!!.isMoved && !withMovement && !delete && !used) {
                if (!entity!!.isMovement) {
                    return entity!!.speed
                }
            } else if (hide && !entity!!.isMovement) {
                hide = false
            }
        }
        return 0
    }

    override fun canUseSkill(): Boolean {
        return !delete
    }

    override fun endSkill() {
        delete = true
        Entity.updateSkills()
    }

    override fun endTurn() {
        delete = true
    }
}